metrics/meter.go

```diff
@@ -161,101 +133,157 @@ func (NilMeter) Snapshot() Meter { return NilMeter{} }
+// Be sure to unregister the meter from the registry once it is of no use to
+// allow for garbage collection.
+func NewLockFreeRegisteredMeter(name string, r Registry) Meter {
+	c := NewLockFreeMeter()
+	if nil == r {
+		r = DefaultRegistry
+	}
+	r.Register(name, c)
+	return c
+}
+
+// NewLockFreeMeter constructs a new LockFreeMeter and launches a goroutine.
+func NewLockFreeMeter() Meter {
+	if !Enabled {
+		return NilMeter{}
+	}
+	return newLockFreeMeter()
+}
+
+// NewLockFreeMeterForced constructs and registers a new StandardMeter
+// and launches a goroutine no matter the global switch is enabled or not.
+// Be sure to unregister the meter from the registry once it is of no use to
+// allow for garbage collection.
+func NewLockFreeMeterForced(name string, r Registry) Meter {
+	c := newLockFreeMeter()
+	if nil == r {
+		r = DefaultRegistry
+	}
+	r.Register(name, c)
+	return c
+}
+
+func NewMeterForced() Meter {
+	return newLockFreeMeter()
+}
+
+// LockFreeMeter is a lock free implementation of a Meter.
+type LockFreeMeter struct {
+	mu          sync.RWMutex
 	snapshot    *MeterSnapshot
 	a1, a5, a15 EWMA
 	startTime   time.Time
-	stopped     bool
+	dataChan    chan int64
+	stopChan    chan interface{}
+	ticker      *time.Ticker
 }
 
-func newStandardMeter() *StandardMeter {
-	return &StandardMeter{
+func newLockFreeMeter() *LockFreeMeter {
+	meter := &LockFreeMeter{
 		snapshot:  &MeterSnapshot{},
 		a1:        NewEWMA1(),
 		a5:        NewEWMA5(),
 		a15:       NewEWMA15(),
 		startTime: time.Now(),
+		dataChan:  make(chan int64, 10),
+		stopChan:  make(chan interface{}),
+		ticker:    time.NewTicker(5e9),
 	}
+	go func() {
+		for {
+			select {
+			case n := <-meter.dataChan:
+				meter.mu.Lock()
+				meter.snapshot.count += n
+				meter.a1.Update(n)
+				meter.a5.Update(n)
+				meter.a15.Update(n)
+				meter.updateSnapshot()
+				meter.mu.Unlock()
+			case <-meter.ticker.C:
+				meter.tick()
+			case <-meter.stopChan:
+				close(meter.dataChan)
+				meter.ticker.Stop()
+				return
+			}
+		}
+	}()
+	return meter
 }
 
-// Stop stops the meter, Mark() will be a no-op if you use it after being stopped.
-func (m *StandardMeter) Stop() {
-	m.lock.Lock()
-	stopped := m.stopped
-	m.stopped = true
-	m.lock.Unlock()
-	if !stopped {
-		arbiter.Lock()
-		delete(arbiter.meters, m)
-		arbiter.Unlock()
-	}
+// Stop stops the meter. If stop was called, Mark becomes a no-op
+func (m *LockFreeMeter) Stop() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	close(m.stopChan)
 }
 
 // Count returns the number of events recorded.
-func (m *StandardMeter) Count() int64 {
-	m.lock.RLock()
+func (m *LockFreeMeter) Count() int64 {
+	m.mu.RLock()
 	count := m.snapshot.count
-	m.lock.RUnlock()
+	m.mu.RUnlock()
 	return count
 }
 
 // Mark records the occurrence of n events.
-func (m *StandardMeter) Mark(n int64) {
-	m.lock.Lock()
-	defer m.lock.Unlock()
-	if m.stopped {
+func (m *LockFreeMeter) Mark(n int64) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
```

Comment:
why is the `Rlock()` needed?

Reply:
I thought because we read from the stopChan, but I'm dumb... We only read from it if it is closed anyway so its not a problem